src/storage/AdminProcessor.h

```diff
@@ -125,13 +190,61 @@ class WaitingForCatchUpDataProcessor : public BaseProcessor<cpp2::AdminExecResp>
         return new WaitingForCatchUpDataProcessor(kvstore);
     }
 
+    ~WaitingForCatchUpDataProcessor() {
+        if (bgThread_) {
+            bgThread_->join();
+        }
+    }
+
     void process(const cpp2::CatchUpDataReq& req) {
-        UNUSED(req);
+        auto spaceId = req.get_space_id();
+        auto partId = req.get_part_id();
+        LOG(INFO) << "Received waiting for catching up data for space "
+                  << spaceId << ", part " << partId;
+        auto ret = kvstore_->part(spaceId, partId);
+        if (!ok(ret)) {
+            onFinished(to(error(ret)));
+            return;
+        }
+        auto part = nebula::value(ret);
+        auto peer = kvstore::NebulaStore::getRaftAddr(HostAddr(req.get_target().get_ip(),
+                                                               req.get_target().get_port()));
+        bgThread_ = std::make_unique<std::thread>([this, part, peer, spaceId, partId] {
+            int retry = FLAGS_waiting_catch_up_retry_times;
+            while (retry-- > 0) {
+                LOG(INFO) << "Waiting for catching up data, peer " << peer
+                          << ", try " << retry << " times";
+                auto res = part->isCatchedUp(peer);
+                switch (res) {
+                    case raftex::AppendLogResult::SUCCEEDED:
+                        onFinished(cpp2::ErrorCode::SUCCEEDED);
```

Comment:
becarefull about that onFinished will delete this, then call ~WaitingForCatchUpDataProcessor and join bgThread_ in the stack of bgThread_, maybe will dead lock that bgThread_ not stopped because join not return.

Reply:
Good catch